[SPARK-25815][k8s] Support kerberos in client mode, keytab-based token renewal.

vanzin opened the pull request:
This change hooks up the k8s backed to the updated HadoopDelegationTokenManager,
so that delegation tokens are also available in client mode, and keytab-based token
renewal is enabled.

The change re-works the k8s feature steps related to kerberos so
that the driver does all the credential management and provides all
the needed information to executors - so nothing needs to be added
to executor pods. This also makes cluster mode behave a lot more
similarly to client mode, since no driver-related config steps are run
in the latter case.

The main two things that don't need to happen in executors anymore are:

- adding the Hadoop config to the executor pods: this is not needed
  since the Spark driver will serialize the Hadoop config and send
  it to executors when running tasks.

- mounting the kerberos config file in the executor pods: this is
  not needed once you remove the above. The Hadoop conf sent by
  the driver with the tasks is already resolved (i.e. has all the
  kerberos names properly defined), so executors do not need access
  to the kerberos realm information anymore.

The change also avoids creating delegation tokens unnecessarily.
This means that they'll only be created if a secret with tokens
was not provided, and if a keytab is not provided. In either of
those cases, the driver code will handle delegation tokens: in
cluster mode by creating a secret and stashing them, in client
mode by using existing mechanisms to send DTs to executors.

One last feature: the change also allows defining a keytab with
a "local:" URI. This is supported in client mode (although that's
the same as not saying "local:"), and in k8s cluster mode. This
allows the keytab to be mounted onto the image from a pre-existing
secret, for example.

Finally, the new code always sets SPARK_USER in the driver and
executor pods. This is in line with how other resource managers
behave: the submitting user reflects which user will access
Hadoop services in the app. (With kerberos, that's overridden
by the logged in user.) That user is unrelated to the OS user
the app is running as inside the containers.

Tested:
- client and cluster mode with kinit
- cluster mode with keytab
- cluster mode with local: keytab
- YARN cluster with keytab (to make sure it isn't broken)

Commit:
[SPARK-25815][k8s] Support kerbers in client mode, keytab-based token renewal.

This change hooks up the k8s backed to the updated HadoopDelegationTokenManager,
so that delegation tokens are also available in client mode.

It also closes the remaining gap in kerberos support which is to support
the current keytab-based token renewal supported by YARN (and partially by
Mesos). That is done by stashing the keytab in a secret on the driver pod,
and then providing it to the HadoopDelegationTokenManager when the
driver starts.

Tested in cluster mode with and without a keytab, and also in client
mode.

SparkQA commented:
**[Test build #98333 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98333/testReport)** for PR 22911 at commit [`7f86e23`](https://github.com/apache/spark/commit/7f86e231922648a9da07ae73026a3a430d2c83f5).

SparkQA commented:
**[Test build #98333 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98333/testReport)** for PR 22911 at commit [`7f86e23`](https://github.com/apache/spark/commit/7f86e231922648a9da07ae73026a3a430d2c83f5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98333/
Test PASSed.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/4680/

mccheah commented:
@ifilonenko @skonto

ifilonenko commented:
Just a note, this PR has some overlaps with https://github.com/apache/spark/pull/22915. But just after a brief skim, I am a bit hesitant on storing keytabs in secrets, but it might have to be done for client mode support.

ifilonenko commented:
> But whether to store it in secrets is not a question. You either store it in a secret or you don't support the keytab/principal feature in Spark at all, and we can delete a bunch of code here.

But can we, for a specific mode, client or cluster, tune whether the keytab is put into a secret? Right now, if a keytab is specified, regardless, a secret is created. I do wish to support keytab/principal so I think that storage in secrets should be fine.

rvesse commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala:
```diff
@@ -126,20 +134,53 @@ private[spark] class KerberosConfDriverFeatureStep(
         hadoopBasedSparkPod))
+
+    if (keytab.isDefined) {
+      val podWitKeytab = new PodBuilder(kerberizedPod.pod)
```
Typo - `Wit` -> `With`

vanzin commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterSchedulerBackend.scala:
```diff
@@ -123,7 +126,11 @@ private[spark] class KubernetesClusterSchedulerBackend(
+  }
+
+  override protected def createTokenManager(): Option[HadoopDelegationTokenManager] = {
+    Some(new HadoopDelegationTokenManager(conf, sc.hadoopConfiguration))
```
I'm not sure I follow your train of thought here so I'll comment on what I understand.

First, the code that creates the secret is in `KerberosConfDriverFeatureStep`. As far as I know, that class is not used in client mode. In client mode the keytab stays in the client machine, with the driver, and the driver just sends DTs to executors. So the whole discussion about secrets is irrelevant in that case.

In cluster mode, you need the driver to have access to the keytab for this feature to work. There are a few ways to achieve that:

- the current YARN mode, which is the keytab lives in the submission host, and is distributed with the application. In k8s this would amount to what I have here: the submission code creates a secret for the driver pod and stashes the keytab in it.

- add the ability to store the keytab in an external place (like HDFS or an HTTP server). That has drawbacks (e.g. people probably wouldn't like that, and there's a chicken & egg problem in HDFS, so you'd still need a kerberos TGT to bootstrap things).

- add a k8s-specific feature of mounting a pre-defined secret in the driver pod. I believe this is what you're suggesting?

I think supporting the first is easy as this change shows, and keeps feature parity with what's already supported in YARN. Unless there's a glaring issue with using secrets that I'm not aware of, I don't see a reason for not doing it.

The third option (pre-defined secret) could also be added. My hope is that you could do it with pre-existing configs (`spark.kubernetes.driver.secrets.` & company), but I don't know how you'd set the `spark.kerberos.keytab` and `spark.kerberos.principal` configs just in the driver - and not in the submission client. So it seems we'd need at least a little bit of code here to support that scenario.

vanzin replied:
Ah, ok I get it now. I can do that. I'll try to include support for (3) but it depends on how much I have to touch other parts of the code. Hopefully not much.

ifilonenko replied:
Yeah, I can always throw up a follow-up for that. No worries

Commit:
Big refactoring of client mode support for kerberos.

This change re-works the k8s feature steps related to kerberos so
that the driver and executor steps are more clearly defined, and
less things need to be done in executor pods.

The main driver behind a lot of these changes is that, in client
mode, no driver-related config steps are run. So all executor-related
steps must be able to be applied even when no driver steps have
previously run and potentially modified the pod or Spark configuration.

The second is that based on the above changes, a bunch of things that
were being done on executor pods are actually not needed. These are:

- adding the Hadoop config to the executor pods: this is not needed
  since the Spark driver will serialize the Hadoop config and send
  it to executors when running tasks.

- mounting the kerberos config file in the executor pods: this is
  not needed once you remove the above. The Hadoop conf sent by
  the driver with the tasks is already resolved (i.e. has all the
  kerberos names properly defined), so executors do not need access
  to the kerberos realm information anymore.

So there is a single thing that executor pods need: mounting the
delegation token secret when one is provided. Everything else is
handled in the driver, regardless of whether it's client or cluster
mode.

The change also avoids creating delegation tokens unnecessarily.
This means that they'll only be created if a secret with tokens
was not provided, and if a keytab is not provided. In either of
those cases, the driver code will handle delegation tokens: in
cluster mode by creating a secret and stashing them, in client
mode by using existing mechanisms to send DTs to executors.

One last feature: the change also allows defining a keytab with
a "local:" URI. This is supported in client mode (although that's
the same as not saying "local:"), and in k8s cluster mode. This
allows the keytab to be mounted onto the image from a pre-existing
secret, for example.

Finally, the new code always sets SPARK_USER in the driver and
executor pods. This is in line with how other resource managers
behave: the submitting user reflects which user will access
Hadoop services in the app. (With kerberos, that's overridden
by the logged in user.) That user is unrelated to the OS user
the app is running as inside the containers.

Tested:
- client and cluster mode with kinit
- cluster mode with keytab
- cluster mode with local: keytab
- YARN cluster with keytab (to make sure it isn't broken)

Commit:
Merge branch 'master' into SPARK-25815

Commit:
Merge branch 'master' into SPARK-25815.tmp

Commit:
Unit tests.

Commit:
Merge branch 'master' into SPARK-25815

vanzin commented:
To address the comments I had to change a whole bunch of things; mainly because I noticed that the client and cluster paths were really different, when they should for the most part be similar.  Check the updated PR description for details.

SparkQA commented:
**[Test build #98832 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98832/testReport)** for PR 22911 at commit [`a129314`](https://github.com/apache/spark/commit/a129314abb45413d7b46e8510b2a2ea7d61fa916).

vanzin commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkPod.scala:
```diff
@@ -18,7 +18,30 @@ package org.apache.spark.deploy.k8s
+   * it can do matching without needing to exhaust all the possibilities. If the function
+   * is not applied, then the original pod will be kept.
+   */
+  def transform(fn: PartialFunction[SparkPod, SparkPod]): SparkPod = fn.lift(this).getOrElse(this)
```
I added this because I started to get tired of code like the following:

```
val someIntermediateName = someOption.map { blah =>
   // create the updated pod
}.getOrElse(previousPodName)

// lather, rinse, repeat
```

To me that's hard to follow and brittle, and this pattern makes things clearer IMO.

ifilonenko replied:
I would think that this change is out of the scope of this PR, but I do love the use of a PartialFunction here. Thanks for this!

Commit:
Fix YARN tests.

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/5029/

SparkQA commented:
**[Test build #98833 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98833/testReport)** for PR 22911 at commit [`d8416cd`](https://github.com/apache/spark/commit/d8416cda0e23af43a86b129a4a102f37bcbdaf4c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5029/
Test PASSed.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/5030/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/5030/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5030/
Test PASSed.

vanzin commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DelegationTokenFeatureStep.scala:
```diff
@@ -0,0 +1,114 @@
+ * When preparing pods, this step will mount the delegation token secret (either pre-defined,
+ * or generated by this step when preparing the driver).
+ */
+private[spark] class DelegationTokenFeatureStep(conf: KubernetesConf[_], isDriver: Boolean)
```
During the recent discussion in the sig meeting I was thinking about this guy. I think it may not be necessary to mount this secret in executors, and let the driver propagate the tokens to executors through its normal means.

e.g. later if/when adding code to monitor delegation tokens for updates, that would mean only the driver has to do it.

I'll git this a try; if it works, it might be feasible to re-merge this code with the kerberos step.

ifilonenko replied:
This is what I meant above when I said that the `HadoopKerberosLogin` logic could be deleted. The assumption here is that the secret should not be created as the keytab will use the HadoopDelegationTokenManager logic. The only secret that should be _created_ would be the keytab. However, I personally thought that we should point to a secretName that is either the delegationToken or the keytab. Hence why I suggested that the secretName and secretItemKey remain.

vanzin replied:
I think you're missing one use case there. There are 3 use cases:

- keytab: keytab is provided to driver, driver handles kerberos login, creates tokens, distributes tokens to executors
- pre-defined token secret: secret is mounted in the driver, env variable is set, driver loads them and distributes to executors
- default kerberos case: submission client generates delegation tokens, creates a secret for them, then this behaves like the bullet above

The third use case is actually the most common. In your reply above you're only covering the other two. My code covers all three.

I'm just saying that this code actually doesn't need to do anything on the executor side, because the driver takes care of everything when the credentials are provided.

Commit:
Let the driver handle all the kerberos needs.

SparkQA commented:
**[Test build #98838 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98838/testReport)** for PR 22911 at commit [`d4a84bc`](https://github.com/apache/spark/commit/d4a84bc32d6e07c91b72d83913a24719e411369a).

SparkQA commented:
**[Test build #98833 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98833/testReport)** for PR 22911 at commit [`d8416cd`](https://github.com/apache/spark/commit/d8416cda0e23af43a86b129a4a102f37bcbdaf4c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98833/
Test PASSed.

ifilonenko commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:
```diff
@@ -144,6 +144,10 @@ private[spark] class BasicExecutorFeatureStep(
         .addToRequests("cpu", executorCpuQuantity)
         .endResources()
+        .addNewEnv()
+          .withName(ENV_SPARK_USER)
```
I see that you noted that this is always done across resource managers. What is the reason for that, just wondering? as I introduced it exclusively in the HadoopSteps

vanzin replied:
If you don't do this, whatever is the OS user in the container will become the identity used to talk to Hadoop services (when kerberos is not on).

In YARN, for example, that would be the "yarn" user.

In k8s, with the current image, that would be "root".

You probably don't want that by default. We're talking about non-secured Hadoop here, so users can easily override this stuff, but by default let's at least try to identify the user correctly.

ifilonenko commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala:
```diff
@@ -69,97 +72,191 @@ private[spark] class KerberosConfDriverFeatureStep(
-        kubernetesConf.krbConfigMapName,
-        fileLocation)
+  // Visible for testing.
+  def createDelegationTokens(): Array[Byte] = {
```
nit: Why isn't this logic done in a `KubernetesDelegationTokenManager` or some Util class? It seems we are over-running this step with a lot of functionality? no? You are also just mocking via
`when(step).createDelegationToken()`, I personally don't agree with the coverage brought up by this method of unit-testing but this is an extension of the discussion on a separate PR. i.e. I would mock the tokenManager and test to ensure that the `obtainDelegationTokens` take in a particular `creds` as an argument (also, so that one could test the if statement below).`

vanzin replied:
> Why isn't this logic done in a KubernetesDelegationTokenManager or some Util class?

Because that's not needed. What exactly would doing that achieve?

> I would mock the tokenManager and test to ensure that the obtainDelegationTokens take in a particular creds

That's not what this code is testing. This code is testing that when delegation tokens are created, they are stashed in a secret. For that I'm mocking the creation of delegation tokens, because without any services to talk to, delegation tokens will not be created.

(BTW, that's what mocks should be used for.)

vanzin replied:
> I'm mocking the creation of delegation tokens

BTW there are other ways to achieve that, e.g. with `UserGroupInformation.createTestUser()` and `doAs`, but I think the current test is easier to understand.

ifilonenko replied:
> Because that's not needed. What exactly would doing that achieve?

I mean that is the function of a DelegationTokenManager, no? To handle nuances related to DT management, both for creation and renewal. _shrug_

> This code is testing that when delegation tokens are created, they are stashed in a secret

Well there is logic being introduced here beyond just `tokenManager.obtainDelegationToken()`, right? And if the purpose of the tests is to take the results from delegation token creation and put into a secret, then it would make sense to mock a `KubernetesDelegtionTokenManager` and just call `kdtm.obtainTokens()`, where this logic would be housed. Else, if we were to keep this logic within this step, i'd prefer that all parts of the logic are mocked. 

Once again, a matter of opinion, that differs between unit-testing strategy. I'll loop in @mccheah

> I think the current test is easier to understand.

I agree, but is it comprehensive is what I am talking about here?

vanzin replied:
> that is the function of a DelegationTokenManager. To handle nuances related to DT management

Yes. And the test here is not interested in whether DT creation works or not. It's interested in that if they are created, then they are stashed in a secret.

> then it would make sense to mock a KubernetesDelegtionTokenManager and just call kdtm.obtainTokens()

Would it make you happier if instead of having instead of the current method, have this:

```
def createDelegationTokens(creds: Credentials)
```

The only extra thing you'd really be testing is the call to `SparkHadoopUtil.get.serialize`. I don't really see a lot of gains there.

vanzin replied:
> Hmm, I somewhat disagree, but maybe I am just overly opinionated.

The code that creates delegation tokens is not in this class. If delegation token creation is to be tested, it needs to be tested in the test code for `HadoopDelegationTokenManager`.

All this class does is take the tokens that other class creates and stash them in a secret.

I'll change this to not need a mock at all, mostly to make a point.

ifilonenko commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala:
```diff
@@ -69,97 +72,191 @@ private[spark] class KerberosConfDriverFeatureStep(
-      hFiles <- hadoopConfigurationFiles
-    } yield {
-      HadoopBootstrapUtil.buildHadoopConfigMap(hName, hFiles)
+    Seq[HasMetadata]() ++ {
```
nit: is there a particular advantage gained by this style of Seq() building? I prefer the .toSeq approach as I personally don't like `if {...} else { Nil }`. Once again, why was the `Util` class removed? Isn't this a lot of logic to throw into this one step?

vanzin replied:
re: the style, same reasoning as the `transform` method I added. It avoids having to create a bunch of local variables and then concatenate them at the end.

re: the Util class, it existed because the methods were being called from multiple feature steps. Now there's only one step that does that stuff, so the util class became unnecessary.

ifilonenko replied:
> the style, same reasoning as the transform method I added. It avoids having to create a bunch of local variables and then concatenate them at the end.

The `transform` I like because it gets rid of: 
```
  val someIntermediateName = someOption.map { blah =>
     // create the updated pod
  }.getOrElse(previousPodName)
```
But that is not what is happening here exactly. Personally, I thought that the previous approach (with local variables + concatonation) was quite clear and didn't really require a change. But its a matter of opinion.

> Now there's only one step that does that stuff, so the util class became unnecessary.

Okay I guess from a re-usability perspective that is true, but imo seems like a lot of logic going into this step. but once again, just a matter of opinion

vanzin replied:
> seems like a lot of logic going into this step

The logic itself isn't really different from before. The only difference is that the "what is done" for each combination of configs now also lives here.

It could be broken down into separate steps, but then you'd have 3 related steps that need to agree on how to treat the configuration of the driver; e.g. if the step that adds they keytab runs, then the step that creates delegation tokens shouldn't run. And that makes the code harder to follow, because you have to look at multiple places to understand how this works.

ifilonenko commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesDriverBuilder.scala:
```diff
@@ -81,14 +81,14 @@ private[spark] class KubernetesDriverBuilder(
-    val maybeHadoopConfigStep =
-      kubernetesConf.hadoopConfSpec.map { _ =>
-        provideHadoopGlobalStep(kubernetesConf)}
+    val otherSteps = Seq(
```
nit: `otherStep` isn't very descriptive to the contents of the steps. I prefer `hadoopSteps` if anything

vanzin replied:
All this code will go away when I get to SPARK-25877, so I'm not super concerned with the names here.

ifilonenko commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStepSuite.scala:
```diff
@@ -0,0 +1,164 @@
+
+  test("create delegation tokens if needed") {
+    val step = spy(createStep(new SparkConf(false)))
+    doReturn(Array[Byte](0x4, 0x2)).when(step).createDelegationTokens()
```
Should we check the contents of the token in the unit test?

vanzin replied:
Seems redundant, but why not.

Commit:
Test dt secret content.

SparkQA commented:
**[Test build #98842 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98842/testReport)** for PR 22911 at commit [`7e9b3a2`](https://github.com/apache/spark/commit/7e9b3a2218359846176136cc98d63eb841d97f15).

Commit:
Don't mock.

SparkQA commented:
**[Test build #98846 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98846/testReport)** for PR 22911 at commit [`78b76a8`](https://github.com/apache/spark/commit/78b76a8ace35b77a6f005dfb26a6e2fd836c1ff4).

SparkQA commented:
**[Test build #98838 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98838/testReport)** for PR 22911 at commit [`d4a84bc`](https://github.com/apache/spark/commit/d4a84bc32d6e07c91b72d83913a24719e411369a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98838/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5032/
Test FAILed.

SparkQA commented:
**[Test build #98842 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98842/testReport)** for PR 22911 at commit [`7e9b3a2`](https://github.com/apache/spark/commit/7e9b3a2218359846176136cc98d63eb841d97f15).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

suxingfate commented:
--keytab is a simpler way IMO to enable long running spark on k8s than DT renew approach

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5033/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5036/
Test FAILed.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #98880 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98880/testReport)** for PR 22911 at commit [`78b76a8`](https://github.com/apache/spark/commit/78b76a8ace35b77a6f005dfb26a6e2fd836c1ff4).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5058/
Test FAILed.

shaneknapp commented:
test this please

SparkQA commented:
**[Test build #98885 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98885/testReport)** for PR 22911 at commit [`78b76a8`](https://github.com/apache/spark/commit/78b76a8ace35b77a6f005dfb26a6e2fd836c1ff4).

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
**[Test build #98880 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98880/testReport)** for PR 22911 at commit [`78b76a8`](https://github.com/apache/spark/commit/78b76a8ace35b77a6f005dfb26a6e2fd836c1ff4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98880/
Test PASSed.

SparkQA commented:
**[Test build #98885 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98885/testReport)** for PR 22911 at commit [`78b76a8`](https://github.com/apache/spark/commit/78b76a8ace35b77a6f005dfb26a6e2fd836c1ff4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98885/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-25815

SparkQA commented:
**[Test build #99404 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99404/testReport)** for PR 22911 at commit [`88f1bb5`](https://github.com/apache/spark/commit/88f1bb516fa6098125fb7f77fd548a1d423cf745).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/5477/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/5477/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5477/
Test PASSed.

SparkQA commented:
**[Test build #99404 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99404/testReport)** for PR 22911 at commit [`88f1bb5`](https://github.com/apache/spark/commit/88f1bb516fa6098125fb7f77fd548a1d423cf745).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99404/
Test FAILed.

Commit:
Merge branch 'master' into SPARK-25815

Commit:
Post-merge fixes.

SparkQA commented:
**[Test build #99627 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99627/testReport)** for PR 22911 at commit [`05333ea`](https://github.com/apache/spark/commit/05333ea11bc50ade8f3ba8e1d2116bad9d3d7c89).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/5684/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/5684/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5684/
Test PASSed.

vanzin commented:
there was a seemingly corrupt xml file in the jenkins worker, I removed it and will retest.

vanzin commented:
retest this please

vanzin commented:
on a non-testing not, any further feedback here?

SparkQA commented:
**[Test build #99630 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99630/testReport)** for PR 22911 at commit [`05333ea`](https://github.com/apache/spark/commit/05333ea11bc50ade8f3ba8e1d2116bad9d3d7c89).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/5686/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/5686/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5686/
Test PASSed.

SparkQA commented:
**[Test build #99630 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99630/testReport)** for PR 22911 at commit [`05333ea`](https://github.com/apache/spark/commit/05333ea11bc50ade8f3ba8e1d2116bad9d3d7c89).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99630/
Test FAILed.

ifilonenko commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStepSuite.scala:
```diff
@@ -0,0 +1,171 @@
+        creds.addSecretKey(new Text("K8S_TEST_KEY"), Array[Byte](0x4, 0x2))
+        testUser.addCredentials(creds)
+
+        val tokens = SparkHadoopUtil.get.serialize(creds)
```
The reason I am particular about using mocks is in the case that the internal logic changes (i.e. we remove using `SparkHadoopUtil` in future iterations, or whatnot.) But tbh, I am not hard-set on this as I see your point, and such changes aren't probable surrounding this code. So I am guess I am fine with it.

ifilonenko reviewed (approved):
This LGTM pending tests passing and a +1 from @mccheah.
I would also love to see this in as soon as possible as the features contained here are extremely desirable.

Commit:
Merge branch 'master' into SPARK-25815

SparkQA commented:
**[Test build #99842 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99842/testReport)** for PR 22911 at commit [`4143603`](https://github.com/apache/spark/commit/414360315a484d7a32e6817ff2197cd1f3f8f43c).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/5866/

SparkQA commented:
**[Test build #99842 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99842/testReport)** for PR 22911 at commit [`4143603`](https://github.com/apache/spark/commit/414360315a484d7a32e6817ff2197cd1f3f8f43c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ArrowCollectSerializer(Serializer):`
  * `class CSVInferSchema(val options: CSVOptions) extends Serializable `
  * `class InterpretedSafeProjection(expressions: Seq[Expression]) extends Projection `
  * `sealed trait DateTimeFormatter `
  * `class Iso8601DateTimeFormatter(`
  * `class LegacyDateTimeFormatter(`
  * `class LegacyFallbackDateTimeFormatter(`
  * `sealed trait DateFormatter `
  * `class Iso8601DateFormatter(`
  * `class LegacyDateFormatter(`
  * `class LegacyFallbackDateFormatter(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99842/
Test FAILed.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #99932 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99932/testReport)** for PR 22911 at commit [`4143603`](https://github.com/apache/spark/commit/414360315a484d7a32e6817ff2197cd1f3f8f43c).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/5938/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/5938/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5938/
Test PASSed.

SparkQA commented:
**[Test build #99932 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99932/testReport)** for PR 22911 at commit [`4143603`](https://github.com/apache/spark/commit/414360315a484d7a32e6817ff2197cd1f3f8f43c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ArrowCollectSerializer(Serializer):`
  * `class CSVInferSchema(val options: CSVOptions) extends Serializable `
  * `class InterpretedSafeProjection(expressions: Seq[Expression]) extends Projection `
  * `sealed trait DateTimeFormatter `
  * `class Iso8601DateTimeFormatter(`
  * `class LegacyDateTimeFormatter(`
  * `class LegacyFallbackDateTimeFormatter(`
  * `sealed trait DateFormatter `
  * `class Iso8601DateFormatter(`
  * `class LegacyDateFormatter(`
  * `class LegacyFallbackDateFormatter(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99932/
Test PASSed.

Commit:
Merge branch 'master' into SPARK-25815

Commit:
Post-merge fixes.

SparkQA commented:
**[Test build #100047 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100047/testReport)** for PR 22911 at commit [`cce3f1d`](https://github.com/apache/spark/commit/cce3f1dd32c204f5c57b3db3a79d49895512f0a5).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/6036/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/6036/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6036/
Test PASSed.

SparkQA commented:
**[Test build #100047 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100047/testReport)** for PR 22911 at commit [`cce3f1d`](https://github.com/apache/spark/commit/cce3f1dd32c204f5c57b3db3a79d49895512f0a5).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100047/
Test FAILed.

vanzin commented:
No more comments here and I plan to push this tomorrow based on the +1s I have.

erictcgs commented:
Hi @vanzin , great work!  I'd just been trying to get this working and was happy to find your PR.  I'm having some trouble getting it working, though - not sure whether an issue in my env or if there's a condition that's not getting caught, hoping you or others could shed some light:

Summary:
- Executors don't seem to be logging in under any scenario using the delegation token
- In client mode, driver connects to hdfs and gets delegation token but the executors don't seem to be using it
- In cluster mode, seems driver pod doesn't get hadoop configs mapped in (this had been working in previous version) so doesn't try using kerberos keytab.  Once I manually add the hadoop configs the driver is able to connect to hdfs, but the executors fail

Environment overview:
- Kerberized HDFS in traditional hadoop cluster
- Launching spark into separate kubernetes cluster, trying to access a parquet dataset

Steps:
1. Cloned your branch, commit cce3f1dd32c204f5c57b3db3a79d49895512f0a5 from today
1. Compiled, created docker images
1. kinit-ed
1. Tried running a small python program to access hdfs, launching in local, cluster, and client modes

**Results - Local mode**

```
./bin/spark-submit \
--master local[4] \
--name spark-hdfs \
--principal user@XXX.COM \
--keytab keytab.kt \
https://xxx.com/spark/hdfs_test.py
```

works, program runs and is able to load parquet and print the first row

**Results - Client mode**

```
./bin/spark-submit \
--master k8s://https://$KUBERNETES_SERVICE_HOST:$KUBERNETES_SERVICE_PORT \
--deploy-mode client \
--conf spark.driver.host=$(getent hosts $HOSTNAME | cut -f1 -d' ') \
<pod-spec-details>
--conf spark.kubernetes.kerberos.krb5.path=/path/krb5.conf \
--conf spark.kerberos.keytab=keytab.kt \
--conf spark.kerberos.principal=user@XXX.COM \
https://xxx.com/spark/hdfs_test.py
```

Driver is able to login to the hdfs server, acquires a delegation token, starts executors

Executors go to connect to HDFS, error out with the following:
```
2018-12-14 00:13:41 WARN  Client:683 - Exception encountered while connecting to the server : org.apache.hadoop.security.AccessControlException: Client cannot authenticate via:[TOKEN, KERBEROS]
2018-12-14 00:13:41 ERROR Executor:95 - Exception in task 0.3 in stage 0.0 (TID 3)
org.apache.spark.SparkException: Exception thrown in awaitResult:
        at org.apache.spark.util.ThreadUtils$.awaitResult(ThreadUtils.scala:226)
        at org.apache.spark.util.ThreadUtils$.parmap(ThreadUtils.scala:290)
...
Caused by: java.io.IOException: Failed on local exception: java.io.IOException: org.apache.hadoop.security.AccessControlException: Client cannot authenticate via:[TOKEN, KERBEROS]; Host Details : local host is: "hdfs-test-1544746404469-exec-1/10.244.42.225"; destination host is: "xxx.com":5555;
        at org.apache.hadoop.net.NetUtils.wrapException(NetUtils.java:776)
        at org.apache.hadoop.ipc.Client.call(Client.java:1480)
        at org.apache.hadoop.ipc.Client.call(Client.java:1413)
```

Seems like they're not trying to connect with the delegation token?

**Results - Cluster mode**

```
./bin/spark-submit \
--master k8s://https://$KUBERNETES_SERVICE_HOST:$KUBERNETES_SERVICE_PORT \
--deploy-mode cluster \
<pod-spec-details>
--conf spark.kubernetes.kerberos.krb5.path=/path/krb5.conf \
--conf spark.kerberos.keytab=keytab.kt \
--conf spark.kerberos.principal=user@XXX.COM \
https://xxx.com/spark/hdfs_test.py
```

Driver and executor pods are launched

Driver tries to connect to hdfs with SIMPLE authentication

Found that hadoop configs weren't being propagated to driver pod, even when adding the `--conf spark.kubernetes.hadoop.configMapName` option

When I manually added the configs and set the driver environment variable HADOOP_CONF_DIR, driver now is able to connect to hdfs and get delegation token, but executors have the same problem observed in client mode

vanzin commented:
I've been running tests with the Spark `HdfsTest` example; maybe pyspark does something different. I'll try that tomorrow.

I haven't really tried the config map route, only `HADOOP_CONF_DIR`.

rvesse commented:
This might be a bug, I've complained in the past that HDFS stuff doesn't seem to work unless you set `HADOOP_CONF_DIR` in the submission clients environments

Commit:
Who's a bad merge?

vanzin commented:
This one might have been just a bad merge on my part. Without integration tests this is hard to catch (and I was having trouble with minikube yesterday).

I'll try out my tests on minikube tomorrow.

SparkQA commented:
**[Test build #100118 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100118/testReport)** for PR 22911 at commit [`ccb3956`](https://github.com/apache/spark/commit/ccb39560298a5e54f144b8ba2a43d950289ccf34).

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/6103/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6103/
Test PASSed.

SparkQA commented:
**[Test build #100118 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100118/testReport)** for PR 22911 at commit [`ccb3956`](https://github.com/apache/spark/commit/ccb39560298a5e54f144b8ba2a43d950289ccf34).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100118/
Test FAILed.

erictcgs commented:
Had a chance to try it - the driver pod is now getting the keytab mounted and is able to generate delegation token, so that's looking good.  I'm still having an issue where the executors seem to not be using the dt and so are getting denied access - are they also supposed to have hadoop_conf_dir set?

erictcgs commented:
@vanzin - how are you compiling spark?  Specifically, what version of hadoop are you testing with?  I'm working with v2.7.7, wondering if there are changes in later versions to the ipc.client and usergroupinformation classes that may handle delegation tokens differently

vanzin commented:
I've tried this again in all configurations I know (and also pyspark) and it's working for me:

- cluster mode: keytab, local: keytab, no keytab
- client mode: keytab / no keytab, client in my laptop, client in k8s pod
- spark-shell and pyspark shell with tgt

I'm building Spark normally. Make sure your images are updated in docker (remember that the pyspark image is *not* build by default anymore).

You should see a log message like this in the executors:

```
2018-12-14 18:52:00 INFO  SparkHadoopUtil:58 - Updating delegation tokens for current user.
```

erictcgs commented:
Update - my setAuthenticationMethod change didn't work - after calling that I see that the executor reports it's trying to use TOKEN method, but it's still getting rejected with an error that it can't login via TOKEN, KERBEROS:
```
18/12/14 23:26:01 DEBUG UserGroupInformation: PrivilegedActionException as:user (auth:TOKEN) cause:org.apache.hadoop.security.AccessControlException: Client cannot authenticate via:[TOKEN, KERBEROS]
18/12/14 23:26:01 DEBUG UserGroupInformation: PrivilegedAction as:user (auth:TOKEN) from:org.apache.hadoop.ipc.Client$Connection.handleSaslConnectionFailure(Client.java:651)
18/12/14 23:26:01 WARN Client: Exception encountered while connecting to the server : org.apache.hadoop.security.AccessControlException: Client cannot authenticate via:[TOKEN, KERBEROS]
```

As a test I tried mounting in my keytab and hadoop_conf_dir into the executors and setting KRB5CCNAME env var (via pod template) - this results in the main user being kerberized, then I commented out the "runassparkuser" section of the coarsescheduler, and now the executors are able to successfully login to HDFS and grab the parquet data.  

It looks like there's some incompatibility in my environment with how the delegation tokens are being used - I think this shouldn't block this merge if it's working for others - is there any other venue people would suggest I continue trying to debug?

vanzin commented:
I suggest filing a bug or starting a mailing list thread. But really I doubt anyone will be able to help if you only provide the errors, without explaining your environment and, especially, the code you're running.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #100257 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100257/testReport)** for PR 22911 at commit [`ccb3956`](https://github.com/apache/spark/commit/ccb39560298a5e54f144b8ba2a43d950289ccf34).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/6229/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/6229/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6229/
Test PASSed.

SparkQA commented:
**[Test build #100257 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100257/testReport)** for PR 22911 at commit [`ccb3956`](https://github.com/apache/spark/commit/ccb39560298a5e54f144b8ba2a43d950289ccf34).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100257/
Test PASSed.

turboFei commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/HadoopConfExecutorFeatureStep.scala:
```diff
@@ -1,40 +0,0 @@
-      pod
-    }
-  }
-}
```
gentle ping @vanzin 

After this pr, seems the HADOOP_CONFIG_MAP is not mounted on executor side.

how does the executor get hadoop conf?

turboFei commented:
> The main two things that don't need to happen in executors anymore are:
> adding the Hadoop config to the executor pods: this is not needed
> since the Spark driver will serialize the Hadoop config and send
> it to executors when running tasks.

gentle ping @vanzin 
seems the executor still need hadoop config.
<img width="1383" alt="image" src="https://github.com/apache/spark/assets/6757692/6632c855-1ab6-47ce-a350-0a344bc02e2e">
